Update email verification items - remove old notices, add banner, update copies and success messages.

Addison-Stavlo opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to # https://github.com/Automattic/wp-calypso/issues/93895 and https://github.com/Automattic/wp-calypso/issues/94247

## Proposed Changes

1. Removes the email not verified notice and email pending change notice that were previously underneath the input and moves these concerns into the input's explanation text, a new banner, and new success message on successful email change. The various forms of the explanation text are shown immediately below:
    * General case (unchanged).
    <img width="486" alt="Screenshot 2024-09-24 at 1 01 49 PM" src="https://github.com/user-attachments/assets/029326e2-8f95-47b2-ba65-fbbd98464c08">
    
    * Email not verified case:
    <img width="505" alt="Screenshot 2024-09-24 at 1 01 23 PM" src="https://github.com/user-attachments/assets/1300c700-0c6e-4ab0-87a7-c4bbf710e6c8">
    
    * Email pending change case:
    <img width="491" alt="Screenshot 2024-09-24 at 1 02 16 PM" src="https://github.com/user-attachments/assets/f4d7e331-c525-4cb4-b61e-77b272c30450">

2. Adds a new email verification banner at the top of /me and /me/account that opens the email verification dialog when clicked.
<img width="850" alt="Screenshot 2024-09-24 at 1 05 07 PM" src="https://github.com/user-attachments/assets/9bfebba0-c346-413f-ac7d-ef4f9985af70">

3. The email verification dialog has been updated to close on route change, and to send an event to signal focus/scroll behavior on the email input if on /me/account. It has also been updated to allow resending an email for pending email change, as it was previously only set up for verification email.
    *  `updated to close on route change` - previously this dialog was not shown /me/account, so clicking the link to /me/account would close the dialog since its parent controller was no longer rendered. Now that we have the new verification banner controlling the dialog which is also shown on /me/account, the parent controller does not dismount and close the dialog. Therefore we ensure it closes on route change.
    * When on /me/account, the link to change the email that would generally lead to /me/account now focuses and scrolls the email input (or cancel pending change button if applicable).
4. Adds a SuccessNotice when the email is updated, mentioning an email was sent to the new address.
    * This has been added on both /me/account as well as /me/security/account-email
    * On /me/account, the success notice copy is conditional to the changes made (other settings only, email only, or both email and other settings).
        * Only other settings changed
        <img width="346" alt="Screenshot 2024-09-24 at 10 49 43 AM" src="https://github.com/user-attachments/assets/d246094d-61b3-45f7-af4b-af7b5ab27ff2">

        * Only email changed
        <img width="600" alt="Screenshot 2024-09-24 at 10 40 50 AM" src="https://github.com/user-attachments/assets/770e3640-a727-46f6-97c1-d6211e241273">

        * Both email and other settings changed
        <img width="600" alt="Screenshot 2024-09-24 at 10 49 30 AM" src="https://github.com/user-attachments/assets/84577018-d648-4160-b737-175fca1fae9c">


5. Adds a success notice when the user is redirected to /me/account after confirming a pending email change. This contains the previous information about updating domain information if necessary and a the same link to updating domains contact info that was previously in the notice.
    * <img width="600" alt="Screenshot 2024-09-23 at 5 15 26 PM" src="https://github.com/user-attachments/assets/c3ed1120-9598-493f-8a22-48f2d33673c7">
    * Also updates the middleware that redirects legacy route `/settings/account` to `/me/account` to include the `?new_email_result=1` param if it exists, and other middleware will check for this param, remove it, and add the corresponding notice.
6. POTENTIAL FOLLOW-UPs
    * Some basic unit tests. We removed some unit tests for the one of the now unused notices that verified it does not show up unless the email is unverified. We should at the very least add a unit test for the banner to ensure it only shows up if unverified or if pending change.
    * We may want to update the banner component to support an `isCritical` or `isWarning` type case. Currently it is set to inherit interface colors (or maybe jetpack colors), and we override styles for our specific use to get the red. If this is intended to be used for more warnings in the future, we should move those styles into the banner component and allow them to be enabled with a prop.
BEFORE
<img width="400" alt="Screenshot 2024-09-20 at 2 50 48 PM" src="https://github.com/user-attachments/assets/b3ecfa21-32e7-4cf4-8d8b-9fa19e5dac6b">

<img width="400" alt="Screenshot 2024-09-18 at 2 03 32 PM" src="https://github.com/user-attachments/assets/d10e8ec2-be60-4f00-99fa-486b6567b5ab">

<img width="400" alt="Screenshot 2024-09-20 at 3 03 11 PM" src="https://github.com/user-attachments/assets/e60d6778-8c41-47b7-926c-2c216b992dd1">

AFTER
As shown above in description items.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* as expressed on issues https://github.com/Automattic/wp-calypso/issues/93895 and https://github.com/Automattic/wp-calypso/issues/94247
* The notices were not prominent enough (this banner approach allows extending them to other pages like /me).
* The notices were piled on redundant in some cases like unverified + email pending change.
* There was no success notice for a successful email change confirmation.

## Testing Instructions

1. Create a couple test email addresses, create a new account on WordPress.com and do not verify the email. I think logging in with google automatically verifies (?), so I used the "continue with email" create account option.
2. Load this branch of calypso, 
3. go to /me/account.
4. Verify the text underneath the input updates to reflect the email is not verified.
5. Verify the new banner appears at the top of /me and /me/account.
6. On /me, verify clicking the banner opens the dialog. Verify the "change" button on the dialog redirects to /me/account and closes the dialog.
    * Bonus - verify the "resend email" button in the dialog still works here and sends an email to your inbox.
8. On /me/account, verify clicking the banner opens the dialog. Verify the "change" button closes the dialog and focus/scrolls the input for use.
9. Verify your email via the link in the email message. 
10. Change your email on /me/account. When you submit you should see a success notice mentioning the email address to check.
   * Later test this along with other settings like changing your primary site, as well as just changing the primary site with no email change. Verify the success notice updates contextually to what has changed as described in the description above.
11. The input should now be locked, and the copy underneath should have changed. Verify there is a link/button to cancel the pending email change.
12. Re-open the dialog from the banner. Click "resend email" and verify you get another copy of the pending change email in your inbox. 
13. Re-open the dialog from the banner. Verify clicking "change" now focuses the cancel pending change button.
14. Re-open the dialog from the banner. 
15. Click the button to cancel the pending change. Verify the input is usable and shows the old email address.
16. Go to `*/settings/account?new_email_result=1`. Verify this redirects to /me/account, they query param is gone, and there is a success notice with a link to domains.
    * We can not test email -> link click to confirm change -> into calypso here easily since these links are handled through and redirected in production. So instead we test where a successful email change would land at `*/settings/account?new_email_result=1`.
17. Verify that once email is verified and there is no pending change, there is no verification banner shown and that the message under the input does not note a need for verification or a pending change.


<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Addison-Stavlo committed:
remove notice under input and update copy

Addison-Stavlo committed:
update success notices

Addison-Stavlo committed:
add fallback for one of those...

Addison-Stavlo committed:
focus input on unlock

Addison-Stavlo committed:
init new banner component

Addison-Stavlo committed:
add notice icon

Addison-Stavlo committed:
move notice out of content frame

Addison-Stavlo committed:
style banner to fit and remove bad unused style rule for mobile

Addison-Stavlo committed:
add banner to account page

Addison-Stavlo committed:
update unlock field button styles

Addison-Stavlo committed:
add note to self

Addison-Stavlo committed:
remove unused component

Addison-Stavlo committed:
unlock and highlight input when change is clicked from modal on the page with that form

Addison-Stavlo committed:
ensure dialog closes on route change

Addison-Stavlo committed:
adjust button and anchor style to be consistent

Addison-Stavlo committed:
allow editing pending email

Addison-Stavlo committed:
allow saving pending email change in security form

Addison-Stavlo committed:
show correct email with change notice on security input

Addison-Stavlo committed:
send success email message from account component

Addison-Stavlo committed:
move emailToShow to component state and control with hooks, fixes issue with input jumping to pending value if you type in old email value.

Addison-Stavlo committed:
fix duplicate notices and only show email when it changed with that submission

Addison-Stavlo committed:
expand explanation text for domains and cancel pending cases

Addison-Stavlo committed:
more commenting around emailSettingToShow state

Addison-Stavlo committed:
remove test file for removed component

Addison-Stavlo committed:
remove input lock and rely on canceling pending change (#94787)

Addison-Stavlo committed:
add notice on confirmed email change

Addison-Stavlo committed:
remove extra domain info on pending change description, the change is still pending

Addison-Stavlo committed:
update copy per design

Addison-Stavlo committed:
consolidate success messages

Addison-Stavlo committed:
add scroll and focus functionality from help dialog

Addison-Stavlo committed:
make the scrollTo smooth

Addison-Stavlo committed:
fix resend email for pending email change

Addison-Stavlo committed:
fix failing gravatar tests

Addison-Stavlo commented on client/blocks/edit-gravatar/test/index.jsx:
```diff
@@ -36,6 +36,7 @@ jest.mock( 'calypso/components/file-picker', () =>
 		</div>
 	) )
 );
+jest.mock( 'calypso/state/selectors/get-user-settings', () => jest.fn( () => {} ) );
```
This simply fixes unit tests. Edit-gravatar uses the dialog which was not previously connected to user settings. Since the user settings state pulls settings off of nested objects, just the state connection of the component caused this test to fail on accessing properties of undefined. This bypasses that error so the edit-gravatar tests can run as expected.

Addison-Stavlo commented on client/components/email-verification/email-verification-dialog/index.jsx:
```diff
@@ -4,25 +4,43 @@ import { get, includes } from 'lodash';
+import isPendingEmailChange from 'calypso/state/selectors/is-pending-email-change';
+import { setUserSetting } from 'calypso/state/user-settings/actions';
+import { saveUnsavedUserSettings } from 'calypso/state/user-settings/thunks';
 
```
this is the file for the dialog/modal. Where we have:

1. updated it to close on route change (reasons noted in description).
2. updated it to handle resending an email for a pending email change (as opposed to only verification).
3. updated it to handle the case of already being on /me/account where the previous anchor led.

Addison-Stavlo commented on client/components/email-verification/index.js:
```diff
@@ -8,6 +9,7 @@ import { successNotice } from 'calypso/state/notices/actions';
 
 export default function emailVerification( context, next ) {
```
This middleware handles the success notice for email verification. We have extended it to handle new email change confirmation as well.

Addison-Stavlo commented on client/me/account/account-email-field.tsx:
```diff
@@ -1,25 +1,20 @@
-import type { ResponseDomain } from 'calypso/lib/domains/types';
 import type { UserSettingsType } from 'calypso/state/selectors/get-user-settings';
 import type { ChangeEvent } from 'react';
 
```
This is the main email change field. We have updated:
* To remove existing notices
* Conditionalize explanation text copies
* Provide functionality for triggering focus and scroll to the input (or button to unlock in pending change cases)

Addison-Stavlo commented on client/me/account/main.jsx:
```diff
@@ -612,6 +612,39 @@ class Account extends Component {
 			return;
 		}
 
+		// Determine success message based on what has been updated.
```
This is for the success message shown when saving changes in /me/account.

Addison-Stavlo commented on client/me/security-account-email/index.tsx:
```diff
@@ -48,9 +48,18 @@ const SecurityAccountEmail = ( { path }: { path: string } ) => {
 				dispatch(
 					successNotice(
-						translate( 'Your account email address was updated successfully.' ),
+						translate(
```
This is for the notice shown when saving the email change on /me/security/account-email

Addison-Stavlo commented on client/me/style.scss:
```diff
@@ -35,12 +35,12 @@ body.is-section-me {
 			.main {
 				padding: 24px;
 				border-block-end: 1px solid var(--studio-gray-0);
+				background: var(--color-surface);
+				border-radius: 8px; /* stylelint-disable-line scales/radii */
+				box-shadow: 0 0 17.4px 0 rgba(0, 0, 0, 0.05);
+				overflow-y: auto;
+				overflow-x: hidden;
```
We move these styles into the main element so the verification banner can be outside the content frame.

DustyReagan replied:
I think this is removing padding on wide displays. Is this a possible regression or is this intentional?

<img width="1730" alt="Screenshot 2024-09-25 at 2 36 09 PM" src="https://github.com/user-attachments/assets/7c02b5a6-05e8-4318-8bb1-d499b8849858">

Addison-Stavlo replied:
Good catch! This was not an intentional change, I assume the main container may have a max-width set and moving the frame to it has now applied that max-width to the frame, showing this padding change you note.

I actually observed this behavior when expanding my browser window and didn't even realize it was a change from production. 😅 I actually like it, but it was not intentional. We will probably want to revert this change, which means we may need to decide on the max-width for the banner.

Currently the banner max-width is set to match that of the content, so at wide resolutions the banner and the content frame currently match width. If we fix this regression, we will have to decide between the banner:

1. Keeping this same max width and not look aligned with the content frame (but will be aligned with the content inside the frame).
2. Removing the max width from the banner and allowing it to stretch as wide as the content frame does... which could feel way too wide in some cases.

1 may make the most sense, but @javierarce if you have thoughts regarding this.

Addison-Stavlo commented on client/me/style.scss:
```diff
@@ -74,11 +74,12 @@ body.is-section-me {
+
+	.email-verification-banner {
+		max-width: 1040px;
+		// Note - may want to move these to the banner component and support a special isCritical case.
```
Potential follow up to do, but I don't want to inflate this PR further.

Addison-Stavlo commented on client/my-sites/site-settings/controller.js:
```diff
@@ -150,6 +151,12 @@ export function legacyRedirects( context, next ) {
 		'connected-apps': '/me/security/connected-applications',
 	};
 
+	if ( section === 'account' && context.query.new_email_result ) {
+		return page.redirect(
+			addQueryArgs( '/me/account', { new_email_result: context.query.new_email_result } )
+		);
+	}
```
This is the legacyRedirects middleware. Currently a confirmed email change redirects to a legacy route which is then re-routed to /me/account. Without this change, it would lose its query param noting that the new email change was a success.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16880746

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @Addison-Stavlo for including a screenshot in the description! This is really helpful for our translators.

javierarce commented:
Thanks for the super detailed testing instructions and you great work on this!

I found a couple of minor things:

> On /me/account, verify clicking the banner opens the dialog. Verify the "change" button closes the dialog and focus/scrolls the input for use.

This worked well, but when I hovered the link the cursor didn't change.

https://github.com/user-attachments/assets/0d4c9063-621e-4620-9e8e-a32600faf675

I also noticed that this particular modal doesn't disspear if I click on the backdrop.

> On /me, verify clicking the banner opens the dialog. Verify the "change" button on the dialog redirects to /me/account and closes the dialog.

Would it be possible to also focus on the email or the cancel link in this case? The switch between pages happens so fast that I think some people could feel a bit lost after the modal is dismissed.

Addison-Stavlo commented:
> I also noticed that this particular modal doesn't disspear if I click on the backdrop.

Yeah, that is a setting on the modal component that this specific implementation of chose to not enable (I haven't changed anything in this PR regarding that). But its an easy change to allow it to close when clicking the backdrop if thats what we want.

Addison-Stavlo committed:
add cursor pointer to the change button

Addison-Stavlo committed:
fix dialog to close on clicking backdrop

Addison-Stavlo committed:
allow scroll/focus from external pages

Addison-Stavlo commented:
Thanks for looking @javierarce and nice finds!

I updated the items you mentioned. The button should show a cursor on hover now, the dialog should close on overlay click, and scroll/focus functionality should work when using the modal from a separate page.

Addison-Stavlo commented on client/me/account/account-email-field.tsx:
```diff
@@ -209,6 +187,37 @@ const AccountEmailField = ( {
+	}, [ isEmailChangePending ] );
+
+	useEffect( () => {
+		// Check for query param used to indicate a need to scroll to the input, this is added when
+		// redirecting to this page for the purpose of email change.
+		const urlQueryParams = new URL( window.location.href )?.searchParams;
+		const focusEmail = urlQueryParams.get( 'focusEmail' );
+		if ( focusEmail ) {
+			scrollAndFocus();
+			window.history.replaceState( {}, '', removeQueryArgs( window.location.href, 'focusEmail' ) );
+		}
+		// Listen for an event signalling to focus and scroll to the email field or button. This
+		// happens when we are already on a page with the input and need to trigger the
+		// functionality.
+		emailFormEventEmitter.addEventListener( 'highlightInput', scrollAndFocus );
```
Instead of using a query param for cases coming from another page, and an event emitter for cases on triggering this on the same page, I think the other option is moving this to Redux but I am conflicted. In both cases this is more of a signal/event to trigger functionality and not related to any state to manage, so Redux seems a little weird but we could probably manage it with a true/false toggle. But that seems like potential to get out-of-sync in some way (?), and generally it feels like extra boilerplate code and overkill for the purpose. 

Id be curious thoughts of other developers reviewing.

DustyReagan replied:
I think the query param implementation makes sense. But does it need the same `isEmailChangePending` conditional check?

Also, I think the scroll only works the first time in Firefox. 🤔 

https://github.com/user-attachments/assets/4f7c0072-825a-413b-a94b-620a5fc37119

Addison-Stavlo replied:
> But does it need the same isEmailChangePending conditional check?

Yeah. If the change is pending the input is disabled, so instead we focus the button that cancels the pending change and unlocks the input. So either way that check is relevant to figure out what element to focus.

Addison-Stavlo replied:
> Also, I think the scroll only works the first time in Firefox. 🤔

Ah interesting! I don't think its just firefox.

So if you open the modal on `/me/account` and click "change" scroll works every time. But if you go to `/me`, open the modal, and click "change", the scroll/focus it only works if you haven't loaded /me/account yet.

<s>I didn't think this would be the case but for whatever reason it seems like the email form component on /me/account doesn't dismount when you go back to /me. So the hook checking for the query param on mount to scroll to the input isn't firing since the component has been mounted and none of its dependencies have changed.</s>

Addison-Stavlo replied:
Ok not component mounting like i suspected, its the component loading before the window url is set in this case.

this should now be resolved with [df256f2](https://github.com/Automattic/wp-calypso/pull/94683/commits/df256f250eeda64b11023d73c04e7a267ecd11be).

DustyReagan reviewed (commented):
This is testing well for me! I left a couple of comments.

Also, I have some feedback on the modal, but this could be suitable for a new issue.

<img width="575" alt="Screenshot 2024-09-25 at 4 12 59 PM" src="https://github.com/user-attachments/assets/0a3a7b2c-9712-4f56-89e3-54af90f97ce3">

I know this is existing copy, but it’s a little confusing. It says to "open the email and click the blue button", but the button in the modal is blue, could it mean that button? 😅 Additionally, if this is a second email change, the verification email is text-based, and there is no button in that email. Finally, the second paragraph in the modal has a smaller font size.

Addison-Stavlo commented:
> Also, I have some feedback on the modal, but this could be suitable for a new issue.

Note the modal itself has been updated separately in https://github.com/Automattic/wp-calypso/pull/94629 and waiting translations.

DustyReagan commented:
> > Also, I have some feedback on the modal, but this could be suitable for a new issue.
> 
> Note the modal itself has been updated separately in #94629 and waiting translations.

Oh! I missed that. Thanks! 😄

Addison-Stavlo committed:
allow scrollTo functionality from /me to work even if /me/account was previously loaded

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16880746).

Addison-Stavlo committed:
more merge conflict resolution

Addison-Stavlo committed:
fix css bracket removed in merge conflict

Addison-Stavlo committed:
fix import order

Addison-Stavlo committed:
fix layout frame regression

Addison-Stavlo commented:
@DustyReagan  - This has been updated to resolve the issues you found, and rebased to include the changes to the dialog modal that hopefully fit a bit better now. 😄 

@javierarce Question for you. [Above Dusty found](https://github.com/Automattic/wp-calypso/pull/94683#discussion_r1775901904) a regression with our layout frame width in larger viewports. I have fixed the regression but it leaves a question for the Banner width.  I have currently kept the banner width the same max-width as the content inside the frame:

![banner-width-to-frame](https://github.com/user-attachments/assets/d4fcb0cf-3762-4a66-94c3-d6e13454d031)

I assume the above may be ideal. The other option is to let the banner expand with the edges of the frame itself which could feel a bit odd at wider widths:

<img width="1124" alt="Screenshot 2024-10-01 at 11 01 58 AM" src="https://github.com/user-attachments/assets/ca45ff87-db0d-4a92-b04b-a852ee840714">

Addison-Stavlo committed:
fix issue with wrong email shown in modal, and href overtaking the banner in some circumstances

DustyReagan reviewed (approved):
Tests well and the changes look good to me! Nice work! 👍

Addison-Stavlo commented:
> Tests well and the changes look good to me! Nice work! 👍

Thanks for reviewing Dusty!

Note that im holding shipping for the new translations, as they don't seem to be fully in place yet:
<img width="888" alt="Screenshot 2024-10-01 at 12 24 21 PM" src="https://github.com/user-attachments/assets/c35f6ce5-64e7-48b1-be9d-fbe40be3d7a9">

Addison-Stavlo commented:
It seems the string freeze check froze at 86% but was actually 100% done. Removing and re-adding the label seemed to update it. 🤷‍♀️
